File: canonical-go/jcs.go
package canonical

import (
	"bytes"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// Value is one JCS value: nil (null), bool, string, Int, []Value, or *Object.
type Value = any

// Int is a canonical decimal integer text (e.g. "-1", "0", "42"). Stored as a
// string rather than a bignum because a validated canonical integer round-trips
// through any integer type as the identity, exactly matching the TS bigint path.
type Int string

// Pair is one key/value entry for constructing an Object.
type Pair struct {
	Key string
	Val Value
}

// Object is an insertion-ordered key/value map; serialization sorts keys by
// UTF-8 byte order.
type Object struct {
	keys []string
	m    map[string]Value
}

// NewObject builds an Object from ordered pairs (no duplicate-key check; callers
// constructing values are assumed not to repeat keys).
func NewObject(pairs ...Pair) *Object {
	o := &Object{m: make(map[string]Value, len(pairs))}
	for _, p := range pairs {
		o.keys = append(o.keys, p.Key)
		o.m[p.Key] = p.Val
	}
	return o
}

func (o *Object) addParsed(key string, v Value) error {
	if _, ok := o.m[key]; ok {
		return noncanonical("JCS_DUPLICATE_KEY", "duplicate key "+strconv.Quote(key))
	}
	o.keys = append(o.keys, key)
	o.m[key] = v
	return nil
}

// Get returns the value for key, if present.
func (o *Object) Get(key string) (Value, bool) {
	v, ok := o.m[key]
	return v, ok
}

// Construction helpers for building values (used by tests / callers).

// P builds a Pair.
func P(k string, v Value) Pair { return Pair{Key: k, Val: v} }

// O builds an Object value from ordered pairs.
func O(pairs ...Pair) Value { return NewObject(pairs...) }

// A builds an array value.
func A(items ...Value) Value {
	if items == nil {
		return []Value{}
	}
	return []Value(items)
}

// IntU builds an integer value from an unsigned 64-bit number.
func IntU(n uint64) Value { return Int(strconv.FormatUint(n, 10)) }

// ============================================================================
// Parser (string -> Value) with duplicate-key detection
// ============================================================================

type jcsParser struct {
	src []rune
	pos int
}

func (p *jcsParser) parse() (Value, error) {
	p.skipWhitespace()
	v, err := p.parseValue()
	if err != nil {
		return nil, err
	}
	p.skipWhitespace()
	if p.pos != len(p.src) {
		return nil, noncanonical("JCS_TRAILING_INPUT", "unexpected character at position "+strconv.Itoa(p.pos))
	}
	return v, nil
}

func (p *jcsParser) parseValue() (Value, error) {
	p.skipWhitespace()
	c, ok := p.peek()
	if !ok {
		return nil, noncanonical("JCS_UNEXPECTED_EOF", "unexpected end of input")
	}
	switch {
	case c == '{':
		return p.parseObject()
	case c == '[':
		return p.parseArray()
	case c == '"':
		return p.parseString()
	case c == 't' || c == 'f':
		return p.parseBool()
	case c == 'n':
		return p.parseNull()
	case c == '-' || (c >= '0' && c <= '9'):
		return p.parseNumber()
	default:
		return nil, noncanonical("JCS_UNEXPECTED_CHAR",
			"unexpected character at position "+strconv.Itoa(p.pos)+": "+string(c))
	}
}

func (p *jcsParser) parseObject() (Value, error) {
	if err := p.expect('{'); err != nil {
		return nil, err
	}
	out := &Object{m: make(map[string]Value)}
	p.skipWhitespace()
	if c, ok := p.peek(); ok && c == '}' {
		p.pos++
		return out, nil
	}
	for {
		p.skipWhitespace()
		if c, ok := p.peek(); !ok || c != '"' {
			return nil, noncanonical("JCS_KEY_NOT_STRING", "expected string key at position "+strconv.Itoa(p.pos))
		}
		key, err := p.parseString()
		if err != nil {
			return nil, err
		}
		p.skipWhitespace()
		if err := p.expect(':'); err != nil {
			return nil, err
		}
		val, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		if err := out.addParsed(key, val); err != nil {
			return nil, err
		}
		p.skipWhitespace()
		c, ok := p.peek()
		switch {
		case ok && c == ',':
			p.pos++
			continue
		case ok && c == '}':
			p.pos++
			return out, nil
		default:
			return nil, noncanonical("JCS_EXPECTED_COMMA_OR_BRACE", "expected ',' or '}' at position "+strconv.Itoa(p.pos))
		}
	}
}

func (p *jcsParser) parseArray() (Value, error) {
	if err := p.expect('['); err != nil {
		return nil, err
	}
	out := []Value{}
	p.skipWhitespace()
	if c, ok := p.peek(); ok && c == ']' {
		p.pos++
		return out, nil
	}
	for {
		val, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		out = append(out, val)
		p.skipWhitespace()
		c, ok := p.peek()
		switch {
		case ok && c == ',':
			p.pos++
			continue
		case ok && c == ']':
			p.pos++
			return out, nil
		default:
			return nil, noncanonical("JCS_EXPECTED_COMMA_OR_BRACKET", "expected ',' or ']' at position "+strconv.Itoa(p.pos))
		}
	}
}

func (p *jcsParser) parseString() (string, error) {
	if err := p.expect('"'); err != nil {
		return "", err
	}
	var sb strings.Builder
	for {
		c, ok := p.peek()
		if !ok {
			return "", noncanonical("JCS_STRING_UNTERMINATED", "unterminated string")
		}
		if c == '"' {
			p.pos++
			return sb.String(), nil
		}
		if c == '\\' {
			p.pos++
			esc, ok := p.peek()
			if !ok {
				return "", noncanonical("JCS_BAD_ESCAPE", "trailing backslash in string")
			}
			p.pos++
			switch esc {
			case '"':
				sb.WriteByte('"')
			case '\\':
				sb.WriteByte('\\')
			case '/':
				sb.WriteByte('/')
			case 'b':
				sb.WriteByte('\b')
			case 'f':
				sb.WriteByte('\f')
			case 'n':
				sb.WriteByte('\n')
			case 'r':
				sb.WriteByte('\r')
			case 't':
				sb.WriteByte('\t')
			case 'u':
				if p.pos+4 > len(p.src) {
					return "", noncanonical("JCS_BAD_UNICODE_ESCAPE", "truncated \\u escape")
				}
				hexStr := string(p.src[p.pos : p.pos+4])
				p.pos += 4
				code, err := strconv.ParseUint(hexStr, 16, 32)
				if err != nil || !isHex4(hexStr) {
					return "", noncanonical("JCS_BAD_UNICODE_ESCAPE", "invalid \\u escape: \\u"+hexStr)
				}
				if code >= 0xd800 && code <= 0xdfff {
					return "", noncanonical("JCS_SURROGATE_ESCAPE",
						"surrogate \\u escape forbidden in canonical JSON: \\u"+hexStr)
				}
				sb.WriteRune(rune(code))
			default:
				return "", noncanonical("JCS_BAD_ESCAPE", "invalid escape \\"+string(esc))
			}
			continue
		}
		if c < 0x20 {
			return "", noncanonical("JCS_CONTROL_IN_STRING",
				fmt.Sprintf("unescaped control character U+%04x", c))
		}
		// Go runes cannot be lone surrogates; raw multi-byte scalars pass through.
		sb.WriteRune(c)
		p.pos++
	}
}

func (p *jcsParser) parseNumber() (Value, error) {
	start := p.pos
	if c, ok := p.peek(); ok && c == '-' {
		p.pos++
	}
	intStart := p.pos
	c, ok := p.peek()
	switch {
	case ok && c == '0':
		p.pos++
	case ok && c >= '1' && c <= '9':
		for {
			d, ok := p.peek()
			if ok && d >= '0' && d <= '9' {
				p.pos++
			} else {
				break
			}
		}
	default:
		return nil, noncanonical("JCS_BAD_NUMBER", "invalid number at position "+strconv.Itoa(start))
	}
	if t, ok := p.peek(); ok {
		if t == '.' {
			return nil, noncanonical("JCS_FRACTIONAL_FORBIDDEN", "fractional numbers are forbidden")
		}
		if t == 'e' || t == 'E' {
			return nil, noncanonical("JCS_EXPONENT_FORBIDDEN", "exponential notation is forbidden")
		}
	}
	text := string(p.src[start:p.pos])
	digitPart := string(p.src[intStart:p.pos])
	if len(digitPart) > 1 && digitPart[0] == '0' {
		return nil, noncanonical("JCS_LEADING_ZERO", "numbers must not have leading zeros: "+text)
	}
	if text == "-0" {
		return nil, noncanonical("JCS_NEGATIVE_ZERO", "'-0' is forbidden")
	}
	return Int(text), nil
}

func (p *jcsParser) parseBool() (Value, error) {
	if p.startsWith("true") {
		p.pos += 4
		return true, nil
	}
	if p.startsWith("false") {
		p.pos += 5
		return false, nil
	}
	return nil, noncanonical("JCS_BAD_LITERAL", "invalid literal at position "+strconv.Itoa(p.pos))
}

func (p *jcsParser) parseNull() (Value, error) {
	if p.startsWith("null") {
		p.pos += 4
		return nil, nil
	}
	return nil, noncanonical("JCS_BAD_LITERAL", "invalid literal at position "+strconv.Itoa(p.pos))
}

func (p *jcsParser) startsWith(lit string) bool {
	r := []rune(lit)
	if p.pos+len(r) > len(p.src) {
		return false
	}
	for i, c := range r {
		if p.src[p.pos+i] != c {
			return false
		}
	}
	return true
}

func (p *jcsParser) skipWhitespace() {
	for {
		c, ok := p.peek()
		if ok && (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
			p.pos++
		} else {
			break
		}
	}
}

func (p *jcsParser) peek() (rune, bool) {
	if p.pos < len(p.src) {
		return p.src[p.pos], true
	}
	return 0, false
}

func (p *jcsParser) expect(c rune) error {
	if cur, ok := p.peek(); ok && cur == c {
		p.pos++
		return nil
	}
	return noncanonical("JCS_EXPECTED_CHAR", "expected '"+string(c)+"' at position "+strconv.Itoa(p.pos))
}

func isHex4(s string) bool {
	if len(s) != 4 {
		return false
	}
	for i := 0; i < 4; i++ {
		c := s[i]
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')) {
			return false
		}
	}
	return true
}

// ============================================================================
// Serializer (Value -> canonical string)
// ============================================================================

// escapeString applies RFC 8785 §3.2.2.2 minimal escaping after NFC. CE §3.2: a
// canonical string MUST NOT begin with U+FEFF (BOM); this applies to every JSON
// string token — keys and values alike — independent of object key count.
// Mid-string U+FEFF (ZWNBSP) is permitted.
func escapeString(s string, sb *strings.Builder) error {
	if strings.HasPrefix(s, "\uFEFF") {
		return noncanonical("UTF8_BOM_FORBIDDEN", "BOM at start of JSON string is forbidden")
	}
	if err := assertAssigned(s); err != nil {
		return err
	}
	n := nfcBytes(s)
	sb.WriteByte('"')
	for _, r := range string(n) {
		switch r {
		case '"':
			sb.WriteString(`\"`)
		case '\\':
			sb.WriteString(`\\`)
		case '\b':
			sb.WriteString(`\b`)
		case '\t':
			sb.WriteString(`\t`)
		case '\n':
			sb.WriteString(`\n`)
		case '\f':
			sb.WriteString(`\f`)
		case '\r':
			sb.WriteString(`\r`)
		default:
			if r < 0x20 {
				sb.WriteString(fmt.Sprintf(`\u%04x`, r))
			} else {
				sb.WriteRune(r)
			}
		}
	}
	sb.WriteByte('"')
	return nil
}

func serialize(v Value, sb *strings.Builder) error {
	switch x := v.(type) {
	case nil:
		sb.WriteString("null")
	case bool:
		if x {
			sb.WriteString("true")
		} else {
			sb.WriteString("false")
		}
	case Int:
		sb.WriteString(string(x))
	case string:
		if err := escapeString(x, sb); err != nil {
			return err
		}
	case []Value:
		sb.WriteByte('[')
		for i, item := range x {
			if i > 0 {
				sb.WriteByte(',')
			}
			if err := serialize(item, sb); err != nil {
				return err
			}
		}
		sb.WriteByte(']')
	case *Object:
		keys := append([]string(nil), x.keys...)
		sort.SliceStable(keys, func(i, j int) bool {
			return bytes.Compare(nfcBytes(keys[i]), nfcBytes(keys[j])) < 0
		})
		sb.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				sb.WriteByte(',')
			}
			if err := escapeString(k, sb); err != nil {
				return err
			}
			sb.WriteByte(':')
			if err := serialize(x.m[k], sb); err != nil {
				return err
			}
		}
		sb.WriteByte('}')
	default:
		return noncanonical("JCS_UNSUPPORTED_TYPE", fmt.Sprintf("unsupported value type %T", v))
	}
	return nil
}

// ============================================================================
// Public API
// ============================================================================

// CanonicalizeString parses a JSON string under the restricted JCS profile and
// returns canonical bytes.
func CanonicalizeString(jsonStr string) ([]byte, error) {
	p := &jcsParser{src: []rune(jsonStr)}
	v, err := p.parse()
	if err != nil {
		return nil, err
	}
	var sb strings.Builder
	if err := serialize(v, &sb); err != nil {
		return nil, err
	}
	return []byte(sb.String()), nil
}

// CanonicalizeValue serializes a typed Value canonically into bytes.
func CanonicalizeValue(v Value) ([]byte, error) {
	var sb strings.Builder
	if err := serialize(v, &sb); err != nil {
		return nil, err
	}
	return []byte(sb.String()), nil
}

// ParseCanonical parses and returns the typed value (for inspection / tests).
func ParseCanonical(jsonStr string) (Value, error) {
	p := &jcsParser{src: []rune(jsonStr)}
	return p.parse()
}

File: canonical-rs/src/jcs.rs
//! Restricted JCS profile per CE v1.3 §4 — parity with `jcs.ts`.
//!
//! Based on RFC 8785 with these domain restrictions:
//!   - Numbers: integers only (decimal, no leading zeros, no `+`, no `-0`, no
//!     exponent). Stored as their validated decimal string — equivalent to the
//!     TS bigint path, since a validated canonical integer round-trips through
//!     bigint as the identity.
//!   - Object keys: no duplicates (checked after unescaping).
//!   - Strings: no surrogate `\u` escapes; no BOM at start.
//!   - Forbidden: NaN, Infinity, fractional, exponential.
//!   - Output: keys sorted by UTF-8 byte order, minimal whitespace.

use std::collections::HashSet;

use unicode_normalization::UnicodeNormalization;

use crate::errors::{CanonicalError, Result};

#[derive(Debug, Clone, PartialEq)]
pub enum JcsValue {
    Null,
    Bool(bool),
    /// Canonical decimal integer text (e.g. "-1", "0", "42").
    Int(String),
    Str(String),
    Array(Vec<JcsValue>),
    /// Insertion-ordered key/value pairs; serialization sorts by UTF-8 bytes.
    Object(Vec<(String, JcsValue)>),
}

impl JcsValue {
    pub fn int_i128(n: i128) -> Self {
        JcsValue::Int(n.to_string())
    }
    pub fn int_u128(n: u128) -> Self {
        JcsValue::Int(n.to_string())
    }
    pub fn string(s: &str) -> Self {
        JcsValue::Str(s.to_string())
    }
    pub fn object(pairs: Vec<(&str, JcsValue)>) -> Self {
        JcsValue::Object(pairs.into_iter().map(|(k, v)| (k.to_string(), v)).collect())
    }
    pub fn array(items: Vec<JcsValue>) -> Self {
        JcsValue::Array(items)
    }

    // ---- read accessors (used to navigate a parsed document, e.g. golden.json) ----

    /// Object field lookup by key. Returns `None` for non-objects or missing keys.
    pub fn get(&self, key: &str) -> Option<&JcsValue> {
        match self {
            JcsValue::Object(pairs) => pairs.iter().find(|(k, _)| k == key).map(|(_, v)| v),
            _ => None,
        }
    }

    /// Borrow the string contents, if this is a `Str`.
    pub fn as_str(&self) -> Option<&str> {
        match self {
            JcsValue::Str(s) => Some(s),
            _ => None,
        }
    }

    /// Borrow the elements, if this is an `Array`.
    pub fn as_array(&self) -> Option<&[JcsValue]> {
        match self {
            JcsValue::Array(items) => Some(items),
            _ => None,
        }
    }
}

fn nc(code: &'static str, msg: impl Into<String>) -> CanonicalError {
    CanonicalError::noncanonical(code, msg)
}

// ============================================================================
// Parser (string -> JcsValue) with duplicate-key detection
// ============================================================================

struct JcsParser {
    chars: Vec<char>,
    pos: usize,
}

impl JcsParser {
    fn new(src: &str) -> Self {
        JcsParser {
            chars: src.chars().collect(),
            pos: 0,
        }
    }

    fn parse(&mut self) -> Result<JcsValue> {
        self.skip_whitespace();
        let value = self.parse_value()?;
        self.skip_whitespace();
        if self.pos != self.chars.len() {
            return Err(nc(
                "JCS_TRAILING_INPUT",
                format!("unexpected character at position {}", self.pos),
            ));
        }
        Ok(value)
    }

    fn parse_value(&mut self) -> Result<JcsValue> {
        self.skip_whitespace();
        let c = match self.peek() {
            Some(c) => c,
            None => return Err(nc("JCS_UNEXPECTED_EOF", "unexpected end of input")),
        };
        match c {
            '{' => self.parse_object(),
            '[' => self.parse_array(),
            '"' => Ok(JcsValue::Str(self.parse_string()?)),
            't' | 'f' => self.parse_bool(),
            'n' => self.parse_null(),
            '-' | '0'..='9' => self.parse_number(),
            _ => Err(nc(
                "JCS_UNEXPECTED_CHAR",
                format!("unexpected character at position {}: {c}", self.pos),
            )),
        }
    }

    fn parse_object(&mut self) -> Result<JcsValue> {
        self.expect('{')?;
        let mut out: Vec<(String, JcsValue)> = Vec::new();
        let mut seen: HashSet<String> = HashSet::new();
        self.skip_whitespace();
        if self.peek() == Some('}') {
            self.pos += 1;
            return Ok(JcsValue::Object(out));
        }
        loop {
            self.skip_whitespace();
            if self.peek() != Some('"') {
                return Err(nc(
                    "JCS_KEY_NOT_STRING",
                    format!("expected string key at position {}", self.pos),
                ));
            }
            let key = self.parse_string()?;
            if !seen.insert(key.clone()) {
                return Err(nc("JCS_DUPLICATE_KEY", format!("duplicate key {key:?}")));
            }
            self.skip_whitespace();
            self.expect(':')?;
            let value = self.parse_value()?;
            out.push((key, value));
            self.skip_whitespace();
            match self.peek() {
                Some(',') => {
                    self.pos += 1;
                    continue;
                }
                Some('}') => {
                    self.pos += 1;
                    return Ok(JcsValue::Object(out));
                }
                _ => {
                    return Err(nc(
                        "JCS_EXPECTED_COMMA_OR_BRACE",
                        format!("expected ',' or '}}' at position {}", self.pos),
                    ))
                }
            }
        }
    }

    fn parse_array(&mut self) -> Result<JcsValue> {
        self.expect('[')?;
        let mut out: Vec<JcsValue> = Vec::new();
        self.skip_whitespace();
        if self.peek() == Some(']') {
            self.pos += 1;
            return Ok(JcsValue::Array(out));
        }
        loop {
            let value = self.parse_value()?;
            out.push(value);
            self.skip_whitespace();
            match self.peek() {
                Some(',') => {
                    self.pos += 1;
                    continue;
                }
                Some(']') => {
                    self.pos += 1;
                    return Ok(JcsValue::Array(out));
                }
                _ => {
                    return Err(nc(
                        "JCS_EXPECTED_COMMA_OR_BRACKET",
                        format!("expected ',' or ']' at position {}", self.pos),
                    ))
                }
            }
        }
    }

    fn parse_string(&mut self) -> Result<String> {
        self.expect('"')?;
        let mut result = String::new();
        loop {
            let c = match self.peek() {
                Some(c) => c,
                None => return Err(nc("JCS_STRING_UNTERMINATED", "unterminated string")),
            };
            if c == '"' {
                self.pos += 1;
                return Ok(result);
            }
            if c == '\\' {
                self.pos += 1;
                let esc = match self.peek() {
                    Some(c) => c,
                    None => return Err(nc("JCS_BAD_ESCAPE", "trailing backslash in string")),
                };
                self.pos += 1;
                match esc {
                    '"' => result.push('"'),
                    '\\' => result.push('\\'),
                    '/' => result.push('/'),
                    'b' => result.push('\u{0008}'),
                    'f' => result.push('\u{000C}'),
                    'n' => result.push('\n'),
                    'r' => result.push('\r'),
                    't' => result.push('\t'),
                    'u' => {
                        if self.pos + 4 > self.chars.len() {
                            return Err(nc("JCS_BAD_UNICODE_ESCAPE", "truncated \\u escape"));
                        }
                        let hex: String = self.chars[self.pos..self.pos + 4].iter().collect();
                        self.pos += 4;
                        let code = u32::from_str_radix(&hex, 16).map_err(|_| {
                            nc("JCS_BAD_UNICODE_ESCAPE", format!("invalid \\u escape: \\u{hex}"))
                        })?;
                        if !hex.bytes().all(|b| b.is_ascii_hexdigit()) {
                            return Err(nc(
                                "JCS_BAD_UNICODE_ESCAPE",
                                format!("invalid \\u escape: \\u{hex}"),
                            ));
                        }
                        if (0xd800..=0xdfff).contains(&code) {
                            return Err(nc(
                                "JCS_SURROGATE_ESCAPE",
                                format!("surrogate \\u escape forbidden in canonical JSON: \\u{hex}"),
                            ));
                        }
                        // code is <= 0xFFFF and not a surrogate => always a valid scalar.
                        result.push(char::from_u32(code).expect("non-surrogate BMP scalar"));
                    }
                    other => {
                        return Err(nc("JCS_BAD_ESCAPE", format!("invalid escape \\{other}")));
                    }
                }
                continue;
            }
            if (c as u32) < 0x20 {
                return Err(nc(
                    "JCS_CONTROL_IN_STRING",
                    format!("unescaped control character U+{:04x}", c as u32),
                ));
            }
            // Rust `char` cannot be a lone surrogate; raw multi-byte scalars pass through.
            result.push(c);
            self.pos += 1;
        }
    }

    fn parse_number(&mut self) -> Result<JcsValue> {
        let start = self.pos;
        if self.peek() == Some('-') {
            self.pos += 1;
        }
        let int_start = self.pos;
        match self.peek() {
            Some('0') => {
                self.pos += 1;
            }
            Some('1'..='9') => {
                while matches!(self.peek(), Some('0'..='9')) {
                    self.pos += 1;
                }
            }
            _ => {
                return Err(nc("JCS_BAD_NUMBER", format!("invalid number at position {start}")));
            }
        }
        match self.peek() {
            Some('.') => return Err(nc("JCS_FRACTIONAL_FORBIDDEN", "fractional numbers are forbidden")),
            Some('e') | Some('E') => {
                return Err(nc("JCS_EXPONENT_FORBIDDEN", "exponential notation is forbidden"))
            }
            _ => {}
        }
        let text: String = self.chars[start..self.pos].iter().collect();
        let digit_part: String = self.chars[int_start..self.pos].iter().collect();
        if digit_part.len() > 1 && digit_part.starts_with('0') {
            return Err(nc("JCS_LEADING_ZERO", format!("numbers must not have leading zeros: {text}")));
        }
        if text == "-0" {
            return Err(nc("JCS_NEGATIVE_ZERO", "'-0' is forbidden"));
        }
        Ok(JcsValue::Int(text))
    }

    fn parse_bool(&mut self) -> Result<JcsValue> {
        if self.starts_with("true") {
            self.pos += 4;
            return Ok(JcsValue::Bool(true));
        }
        if self.starts_with("false") {
            self.pos += 5;
            return Ok(JcsValue::Bool(false));
        }
        Err(nc("JCS_BAD_LITERAL", format!("invalid literal at position {}", self.pos)))
    }

    fn parse_null(&mut self) -> Result<JcsValue> {
        if self.starts_with("null") {
            self.pos += 4;
            return Ok(JcsValue::Null);
        }
        Err(nc("JCS_BAD_LITERAL", format!("invalid literal at position {}", self.pos)))
    }

    fn starts_with(&self, literal: &str) -> bool {
        let lit: Vec<char> = literal.chars().collect();
        if self.pos + lit.len() > self.chars.len() {
            return false;
        }
        self.chars[self.pos..self.pos + lit.len()] == lit[..]
    }

    fn skip_whitespace(&mut self) {
        while let Some(c) = self.peek() {
            if matches!(c, ' ' | '\t' | '\n' | '\r') {
                self.pos += 1;
            } else {
                break;
            }
        }
    }

    fn peek(&self) -> Option<char> {
        self.chars.get(self.pos).copied()
    }

    fn expect(&mut self, c: char) -> Result<()> {
        if self.peek() == Some(c) {
            self.pos += 1;
            Ok(())
        } else {
            Err(nc(
                "JCS_EXPECTED_CHAR",
                format!(
                    "expected '{c}' at position {}, got '{}'",
                    self.pos,
                    self.peek().map(|c| c.to_string()).unwrap_or_else(|| "<EOF>".into())
                ),
            ))
        }
    }
}

// ============================================================================
// Serializer (JcsValue -> canonical string)
// ============================================================================

/// RFC 8785 §3.2.2.2 minimal string escaping, after NFC normalization.
///
/// CE §3.2: a canonical string MUST NOT begin with U+FEFF (BOM). This applies
/// to every JSON string token — keys and values alike — independent of object
/// key count. Mid-string U+FEFF (ZWNBSP) is permitted.
fn escape_string(s: &str, out: &mut String) -> Result<()> {
    if s.starts_with('\u{FEFF}') {
        return Err(nc(
            "UTF8_BOM_FORBIDDEN",
            "BOM at start of JSON string is forbidden",
        ));
    }
    crate::strings::assert_assigned(s)?;
    let normalized: String = s.nfc().collect();
    out.push('"');
    for c in normalized.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{0008}' => out.push_str("\\b"),
            '\t' => out.push_str("\\t"),
            '\n' => out.push_str("\\n"),
            '\u{000C}' => out.push_str("\\f"),
            '\r' => out.push_str("\\r"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    Ok(())
}

fn serialize(value: &JcsValue, out: &mut String) -> Result<()> {
    match value {
        JcsValue::Null => out.push_str("null"),
        JcsValue::Bool(true) => out.push_str("true"),
        JcsValue::Bool(false) => out.push_str("false"),
        JcsValue::Int(s) => out.push_str(s),
        JcsValue::Str(s) => escape_string(s, out)?,
        JcsValue::Array(items) => {
            out.push('[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                serialize(item, out)?;
            }
            out.push(']');
        }
        JcsValue::Object(pairs) => {
            // Sort keys by NFC UTF-8 byte order. Pure ordering — the leading-BOM
            // rejection is enforced by escape_string below (run for every key),
            // so it no longer depends on key count.
            let mut sorted: Vec<&(String, JcsValue)> = pairs.iter().collect();
            sorted.sort_by(|a, b| {
                let ab: Vec<u8> = a.0.nfc().collect::<String>().into_bytes();
                let bb: Vec<u8> = b.0.nfc().collect::<String>().into_bytes();
                ab.cmp(&bb)
            });
            out.push('{');
            for (i, (k, v)) in sorted.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                escape_string(k, out)?;
                out.push(':');
                serialize(v, out)?;
            }
            out.push('}');
        }
    }
    Ok(())
}

// ============================================================================
// Public API
// ============================================================================

/// Parse a JSON string with restricted-JCS validation and return canonical bytes.
pub fn canonicalize_string(json: &str) -> Result<Vec<u8>> {
    let parsed = JcsParser::new(json).parse()?;
    let mut out = String::new();
    serialize(&parsed, &mut out)?;
    Ok(out.into_bytes())
}

/// Serialize a typed `JcsValue` canonically into bytes.
pub fn canonicalize_value(value: &JcsValue) -> Result<Vec<u8>> {
    let mut out = String::new();
    serialize(value, &mut out)?;
    Ok(out.into_bytes())
}

/// Parse and return the typed value (for inspection / round-trip tests).
pub fn parse_canonical(json: &str) -> Result<JcsValue> {
    JcsParser::new(json).parse()
}
